contracts/contracts/coordination/Coordinator.sol

```diff
@@ -424,26 +425,101 @@ contract Coordinator is Initializable, AccessControlDefaultAdminRulesUpgradeable
         return ritual.publicKey;
     }
 
-    function getParticipantFromProvider(
+    function findParticipant(
         Ritual storage ritual,
         address provider
-    ) internal view returns (Participant storage) {
+    ) internal view returns (bool, uint256, Participant storage participant) {
         uint256 length = ritual.participant.length;
-        // TODO: Improve with binary search
-        for (uint256 i = 0; i < length; i++) {
-            Participant storage participant = ritual.participant[i];
+        if (length == 0) {
+            return (false, 0, sentinelParticipant);
+        }
+        uint256 low = 0;
+        uint256 high = length - 1;
+        while (low <= high) {
+            uint256 mid = (low + high) / 2;
+            Participant storage participant = ritual.participant[mid];
             if (participant.provider == provider) {
-                return participant;
+                return (true, mid, participant);
+            } else if (participant.provider < provider) {
+                low = mid + 1;
+            } else if (high == 0) {
+                // prevent underflow of unsigned int
+                return (false, 0, sentinelParticipant);
+            } else {
+                high = mid - 1;
             }
         }
-        revert("Participant not part of ritual");
+        return (false, 0, sentinelParticipant);
+    }
+
+    function getParticipant(
+        Ritual storage ritual,
+        address provider
+    ) internal view returns (Participant storage) {
+        (bool found, , Participant storage participant) = findParticipant(ritual, provider);
+        require(found, "Participant not part of ritual");
+        return participant;
+    }
+
+    function getParticipant(
+        uint32 ritualId,
+        address provider,
+        bool transcript
+    ) public view returns (Participant memory, uint256) {
```

Comment:
this function doesn't seem to be used internally, right? In that case, it could be set to external.

Reply:
✔️